From d49095b5f13163d92b3a1603ed6f06fb1cdc029c Mon Sep 17 00:00:00 2001
From: Wadim Egorov <w.egorov@phytec.de>
Date: Thu, 22 Jun 2017 11:23:37 +0200
Subject: [PATCH] Add createDisplay for QEglFSKmsGbmIntegration

createDisplay will use eglGetPlatformDisplayEXT() to retrieve a EGLDisplay for
the GBM platform.
If eglGetPlatformDisplayEXT is not defined, eglGetDisplay() will be used.

Signed-off-by: Wadim Egorov <w.egorov@phytec.de>
---
 .../eglfs_kms/qeglfskmsgbmintegration.cpp          | 29 ++++++++++++++++++++++
 .../eglfs_kms/qeglfskmsgbmintegration.h            |  1 +
 .../eglfs_kms_support/qeglfskmsscreen.cpp          |  2 +-
 3 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.cpp b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.cpp
index 38419a5..d06d19b 100644
--- a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.cpp
+++ b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.cpp
@@ -57,6 +57,7 @@
 #include <xf86drm.h>
 #include <xf86drmMode.h>
 #include <gbm.h>
+#include <EGL/eglext.h>
 
 QT_BEGIN_NAMESPACE
 
@@ -66,6 +67,34 @@ QEglFSKmsGbmIntegration::QEglFSKmsGbmIntegration()
     : QEglFSKmsIntegration()
 {}
 
+EGLDisplay QEglFSKmsGbmIntegration::createDisplay(EGLNativeDisplayType nativeDisplay)
+{
+    qCDebug(qLcEglfsKmsDebug, "Creating display");
+
+    EGLDisplay display;
+    PFNEGLGETPLATFORMDISPLAYEXTPROC eglGetPlatformDisplayEXT = Q_NULLPTR;
+    eglGetPlatformDisplayEXT =
+        (PFNEGLGETPLATFORMDISPLAYEXTPROC)eglGetProcAddress("eglGetPlatformDisplayEXT");
+    if (eglGetPlatformDisplayEXT) {
+        display = eglGetPlatformDisplayEXT(EGL_PLATFORM_GBM_KHR, nativeDisplay, Q_NULLPTR);
+    } else {
+        qWarning("eglGetPlatformDisplayEXT not available, falling back to legacy path!");
+        display = eglGetDisplay(nativeDisplay);
+    }
+
+    if (Q_UNLIKELY(display == EGL_NO_DISPLAY))
+        qFatal("Could not get EGL display");
+
+    EGLint major, minor;
+    if (Q_UNLIKELY(!eglInitialize(display, &major, &minor)))
+        qFatal("Could not initialize egl display");
+
+    if (Q_UNLIKELY(!eglBindAPI(EGL_OPENGL_ES_API)))
+        qFatal("Failed to bind EGL_OPENGL_ES_API\n");
+
+    return display;
+}
+
 EGLNativeWindowType QEglFSKmsGbmIntegration::createNativeWindow(QPlatformWindow *platformWindow,
                                                      const QSize &size,
                                                      const QSurfaceFormat &format)
diff --git a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.h b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.h
index 727571d..0489dc9 100644
--- a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.h
+++ b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.h
@@ -54,6 +54,7 @@ class QEglFSKmsGbmIntegration : public QEglFSKmsIntegration
 {
 public:
     QEglFSKmsGbmIntegration();
+    EGLDisplay createDisplay(EGLNativeDisplayType nativeDisplay) Q_DECL_OVERRIDE;
 
     EGLNativeWindowType createNativeWindow(QPlatformWindow *platformWindow,
                                            const QSize &size,
diff --git a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_support/qeglfskmsscreen.cpp b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_support/qeglfskmsscreen.cpp
index 4021609..5cc8f79 100644
--- a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_support/qeglfskmsscreen.cpp
+++ b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_support/qeglfskmsscreen.cpp
@@ -72,7 +72,7 @@ private:
 QEglFSKmsScreen::QEglFSKmsScreen(QEglFSKmsIntegration *integration,
                                  QEglFSKmsDevice *device,
                                  QEglFSKmsOutput output)
-    : QEglFSScreen(eglGetDisplay(device->nativeDisplay()))
+    : QEglFSScreen(static_cast<QEglFSIntegration *>(QGuiApplicationPrivate::platformIntegration())->display())
     , m_integration(integration)
     , m_device(device)
     , m_output(output)
-- 
1.9.1

